packages/workflow-designer/src/react/workflow-designer-context.tsx

```diff
@@ -132,6 +146,84 @@ export function WorkflowDesignerProvider({
 		[setAndSaveWorkspace],
 	);
 
+	const copyNode = useCallback(
+		async (
+			sourceNode: Node,
+			options?: {
+				ui?: NodeUIState;
+				connectionCloneStrategy?: ConnectionCloneStrategy;
+			},
+		): Promise<Node | undefined> => {
+			const newNodeDefinition = workflowDesignerRef.current.copyNode(
+				sourceNode,
+				options,
+			);
+
+			if (!newNodeDefinition) {
+				return undefined;
+			}
+
+			const finalNewNode = newNodeDefinition;
+
+			if (
+				finalNewNode.type === "variable" &&
+				finalNewNode.content.type === "file" &&
+				sourceNode.type === "variable" &&
+				sourceNode.content.type === "file"
+			) {
+				const newFileNode = finalNewNode as FileNode;
+
+				const fileCopyPromises = newFileNode.content.files.map(
+					async (fileDataWithOriginalId) => {
+						const tempFileData = fileDataWithOriginalId as FileData & {
+							originalFileIdForCopy: FileId;
+						};
+						const { originalFileIdForCopy, ...newFileData } = tempFileData;
+
+						if (originalFileIdForCopy) {
+							try {
+								await client.copyFile({
+									workspaceId: data.id,
+									sourceFileId: originalFileIdForCopy,
+									destinationFileId: newFileData.id,
+								});
+
+								return newFileData as FileData;
+							} catch (error) {
+								console.error(
+									`Context: Failed to copy file for new fileId ${newFileData.id} (source: ${originalFileIdForCopy}):`,
+									error,
+								);
+
+								return {
+									...newFileData,
+									status: "failed",
+									errorMessage:
+										error instanceof Error ? error.message : "Unknown error",
+								} as FailedFileData;
+							}
+						}
+
+						return newFileData as FileData;
+					},
+				);
+
+				const resolvedFiles = await Promise.all(fileCopyPromises);
+				const newContentForNode: FileContent = {
+					...newFileNode.content,
+					files: resolvedFiles,
+				};
+
```

Comment:
Since there is no feedback and nodes are not duplicated until the file copy is complete, which makes it difficult to know if it's working properly, I want to duplicate the nodes (display them in the UI) before copying the files, and then execute the file copy